### PR TITLE
nmstate: add interface vlan with IP

### PR DESCRIPTION
**Pull request by gkopels**
Add nsmstate interface VLAN with IPv4 and IPv6

### CODE REVIEW

File: pkg/nmstate/policy_test.go [hunk 1]
```diff
@@ -379,6 +380,94 @@ func TestPolicyWithVlanInterface(t *testing.T) {
...
+			sriovInterface:    "ens1",
+			vlanID:            4099,
+			ipv4:              "10.10.10.10",
+			ipv6:              "",
+		},
+	}
+	for _, testCase := range testCases {
+		testPolicy := testCase.testNMStatePolicy.WithVlanInterfaceIP(testCase.sriovInterface, testCase.vlanID,
+			testCase.ipv4, testCase.ipv6)
+		assert.Equal(t, testPolicy.errorMsg, testCase.expectedError)
```
Comment (klaskosk) [inline pkg/nmstate/policy_test.go:436]: make sure to put the expected first, it'll make any error message if the test fails more clear
```suggestion
		assert.Equal(t, testCase.expectedError, testPolicy.errorMsg)
```

File: pkg/nmstate/policy_test.go [hunk 2]
```diff
@@ -379,6 +380,94 @@ func TestPolicyWithVlanInterface(t *testing.T) {
...
+
+		desireState := &DesiredState{}
+		if testCase.expectedError == "" {
+			_ = yaml.Unmarshal(testPolicy.Definition.Spec.DesiredState.Raw, desireState)
+			assert.Equal(t, desireState, &DesiredState{
```
Comment (klaskosk) [inline pkg/nmstate/policy_test.go:441]: same as above, expected should go first

File: pkg/nmstate/policy.go [hunk 3]
```diff
@@ -320,6 +321,71 @@ func (builder *PolicyBuilder) WithVlanInterface(baseInterface string, vlanID uin
...
+	}
+
+	glog.V(100).Infof("Creating NodeNetworkConfigurationPolicy %s with VLAN interface %s and vlanID %d",
+		builder.Definition.Name, baseInterface, vlanID)
+
+	if baseInterface == "" {
+		glog.V(100).Infof("The baseInterface can not be empty string")
+
+		builder.errorMsg = "nodenetworkconfigurationpolicy 'baseInterface' cannot be empty"
+	}
```
Comment (klaskosk) [inline pkg/nmstate/policy.go:338]:
```suggestion
	if baseInterface == "" {
		glog.V(100).Infof("The baseInterface can not be empty string")

		builder.errorMsg = "nodenetworkconfigurationpolicy 'baseInterface' cannot be empty"
		
		return builder
	}
```
For new code we should follow this pattern and then we don't need the
```go
if builder.errorMsg != "" {
        return builder
}
```

File: pkg/nmstate/policy.go [hunk 4]
```diff
@@ -320,6 +321,71 @@ func (builder *PolicyBuilder) WithVlanInterface(baseInterface string, vlanID uin
...
+
+	if net.ParseIP(ipv6Addresses) == nil {
+		glog.V(100).Infof("the vlanInterface contains an invalid ipv6 address")
+
+		builder.errorMsg = "vlanInterfaceIP 'ipv6Addresses' is an invalid ipv6 address"
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
```
Comment (klaskosk) [inline pkg/nmstate/policy.go:358]: can be removed once all the if statements are changed based on the above comment

**Review by klaskosk — COMMENTED** [review]
A few minor comments but otherwise looks pretty good

File: pkg/nmstate/policy_test.go (hunk 1 above)
Comment (gkopels) [reply pkg/nmstate/policy_test.go:436]: changed

File: pkg/nmstate/policy_test.go (hunk 2 above)
Comment (gkopels) [reply pkg/nmstate/policy_test.go:441]: Not 100% sure on this one?

File: pkg/nmstate/policy.go (hunk 3 above)
Comment (gkopels) [reply pkg/nmstate/policy.go]: removed

File: pkg/nmstate/policy.go (hunk 4 above)
Comment (gkopels) [reply pkg/nmstate/policy.go]: removed

**Review by klaskosk — APPROVED** [review]
/lgtm

File: pkg/nmstate/policy.go [hunk 5]
```diff
@@ -320,6 +321,77 @@ func (builder *PolicyBuilder) WithVlanInterface(baseInterface string, vlanID uin
 	return builder.withInterface(newInterface)
 }
 
+// WithVlanInterfaceIP adds VLAN interface with IP configuration to the NodeNetworkConfigurationPolicy.
+func (builder *PolicyBuilder) WithVlanInterfaceIP(baseInterface string, vlanID uint16,
```
Comment (kononovn) [inline pkg/nmstate/policy.go:325]: Function's argument of the same type should be placed next to each other.

**Review by kononovn — CHANGES_REQUESTED** [review]
Please fix 1 comment

File: pkg/nmstate/policy.go (hunk 5 above)
Comment (gkopels) [reply pkg/nmstate/policy.go]: Hi Nikita fixed your comment. Thanks

**Review by klaskosk — APPROVED** [review]

**Review by kononovn — APPROVED** [review]
/lgtm

**Pull request merged**